File: 0415_add_strings/add_strings.cpp
#include <iostream>
#include <string>

using namespace std;

// Time: O(m + n)
// Space: O(n)
string add_strings(string num1, string num2)
{
    int i = num1.size() - 1, j = num2.size() - 1, sum = 0;
    string res;
    while (i >= 0 || j >= 0 || sum)
    {
        if (i >= 0)
            sum += num1[i--] - '0';
        if (j >= 0)
            sum += num2[j--] - '0';
        res.insert(0, 1, (sum % 10) + '0');
        sum /= 10;
    }
    return res;
}

File: 0415_add_strings/add_strings.java
import java.lang.StringBuilder;

public class Solution {
    public String addStrings(String num1, String num2) {
        StringBuilder sb = new StringBuilder();
        int i = num1.length() - 1, j = num2.length() - 1;
        int carry = 0;
        while (i >= 0 || j >= 0) {
            int f = i >= 0 ? num1.charAt(i) - '0' : 0;
            int s = j >= 0 ? num2.charAt(j) - '0' : 0;
            int sum = f + s + carry;
            sb.append(sum % 10);
            carry = sum / 10;
            i--;
            j--;
        }
        if (carry > 0) {
            sb.append(carry);
        }
        return sb.reverse().toString();
    }
}
